src/lib/runtime_config/runtime_config.ml

```diff
@@ -0,0 +1,281 @@
+        in
+        let%map hash =
+          match base with
+          | Hash hash ->
+              return (Some hash)
+          | _ ->
+              Option.value ~default:(return None)
+              @@ List.find_map l ~f:(fun (fld, value) ->
+                     if String.equal "hash" fld then
+                       match value with
+                       | `String hash ->
+                           Some (return (Some hash))
+                       | _ ->
+                           Some
+                             (Error
+                                "Runtime_config.Ledger.of_yojson: Expected \
+                                 the field 'hash' to contain a string")
+                     else None )
+        in
+        {base; num_accounts; hash}
+    | _ ->
+        Error "Runtime_config.Ledger.of_yojson: Expected a JSON object"
+end
+
+module Proof_keys = struct
+  type t = unit
+
+  let to_yojson () = `Assoc []
+
+  let of_yojson = function
+    | `Assoc _ ->
+        Ok ()
+    | _ ->
+        Error "Runtime_config.Proof_keys.of_yojson: Expected a JSON object"
+
+  let combine () () = ()
+end
+
+module Genesis = struct
+  type t = {k: int option [@default None]; delta: int option [@default None]}
+  [@@deriving yojson]
+
+  let of_yojson json =
+    of_yojson @@ yojson_strip_fields ~fields:[|"k"; "delta"|] json
+
+  let combine t1 t2 =
+    { k= opt_fallthrough ~default:t1.k t2.k
+    ; delta= opt_fallthrough ~default:t1.delta t2.delta }
+end
+
+module Daemon = struct
+  type t =
+    { txpool_max_size: int option [@default None]
+    ; genesis_state_timestamp: string option [@default None] }
+  [@@deriving yojson]
+
+  let of_yojson json =
+    of_yojson
+    @@ yojson_strip_fields
+         ~fields:[|"txpool_max_size"; "genesis_state_timestamp"|]
+         json
+
+  let combine t1 t2 =
+    { txpool_max_size=
+        opt_fallthrough ~default:t1.txpool_max_size t2.txpool_max_size
+    ; genesis_state_timestamp=
+        opt_fallthrough ~default:t1.genesis_state_timestamp
+          t2.genesis_state_timestamp }
+end
+
+(** JSON representation:
+
+  { "daemon":
+      { "txpool_max_size": 1
+      , "genesis_state_timestamp": "2000-00-00 12:00:00+0100" }
+  , "genesis": { "k": 1, "delta": 1 }
+  , "proof": { }
+  , "ledger":
+      { "name": "release"
+      , "accounts":
+          [ { pk: "public_key"
+            , sk: "secret_key"
+            , balance: "0.000600000"
+            , delegate: "public_key" }
+          , { pk: "public_key"
+            , sk: "secret_key"
+            , balance: "0.000000000"
+            , delegate: "public_key" } ]
+      , "hash": "root_hash"
+      , "num_accounts": 10 } }
+
+  All fields are optional *except*:
+  * each account in [ledger.accounts] must have a [balance] field
+  * if [ledger] is present, it must feature one of [name], [accounts] or [hash].
+
+*)
+
+type t =
+  { daemon: Daemon.t option [@default None]
+  ; genesis: Genesis.t option [@default None]
```

Comment:
Do these have to be two separate fields? changing timestamp/k/delta requires proofs to be generated if they are not cached

Reply:
I moved the `genesis_state_timestamp` to `genesis`.

If the aim is to keep all of the configuration in a single configuration file, then I feel it's clearer to keep these divisions. Especially so with the `daemon` field expanded to support the contents of `daemon.json`, and as `proof_level` and other 'proof keys' constants are exposed as part of the `proof` field.